File: app/src/main/java/com/elishaazaria/sayboard/ime/MySpeechService.java
/*
 * org.vosk.SpeechService, extended to support other recognizers.
 */

package com.elishaazaria.sayboard.ime;

import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.RequiresPermission;

import com.elishaazaria.sayboard.ime.recognizers.Recognizer;

import java.io.IOException;

import org.vosk.android.RecognitionListener;

public class MySpeechService {
    private final Recognizer recognizer;
    private final int sampleRate;
    private static final float BUFFER_SIZE_SECONDS = 0.2F;
    private final int bufferSize;
    private final AudioRecord recorder;
    private MySpeechService.RecognizerThread recognizerThread;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    @RequiresPermission(android.Manifest.permission.RECORD_AUDIO)
    public MySpeechService(Recognizer recognizer, float sampleRate) throws IOException {
        this.recognizer = recognizer;
        this.sampleRate = (int) sampleRate;
        this.bufferSize = Math.round((float) this.sampleRate * BUFFER_SIZE_SECONDS);
        this.recorder = new AudioRecord(MediaRecorder.AudioSource.VOICE_RECOGNITION, this.sampleRate, 16, 2, this.bufferSize * 2);
        if (this.recorder.getState() == 0) {
            this.recorder.release();
            throw new IOException("Failed to initialize recorder. Microphone might be already in use.");
        }
    }

    public boolean startListening(RecognitionListener listener) {
        if (null != this.recognizerThread) {
            return false;
        } else {
            this.recognizerThread = new MySpeechService.RecognizerThread(listener);
            this.recognizerThread.start();
            return true;
        }
    }

    public boolean startListening(RecognitionListener listener, int timeout) {
        if (null != this.recognizerThread) {
            return false;
        } else {
            this.recognizerThread = new MySpeechService.RecognizerThread(listener, timeout);
            this.recognizerThread.start();
            return true;
        }
    }

    private boolean stopRecognizerThread() {
        if (null == this.recognizerThread) {
            return false;
        } else {
            try {
                this.recognizerThread.interrupt();
                this.recognizerThread.join();
            } catch (InterruptedException var2) {
                Thread.currentThread().interrupt();
            }

            this.recognizerThread = null;
            return true;
        }
    }

    public boolean stop() {
        return this.stopRecognizerThread();
    }

    public boolean cancel() {
        if (this.recognizerThread != null) {
            this.recognizerThread.setPause(true);
        }

        return this.stopRecognizerThread();
    }

    public void shutdown() {
        this.recorder.release();
    }

    public void setPause(boolean paused) {
        if (this.recognizerThread != null) {
            this.recognizerThread.setPause(paused);
        }

    }

    public void reset() {
        if (this.recognizerThread != null) {
            this.recognizerThread.reset();
        }

    }

    private final class RecognizerThread extends Thread {
        private int remainingSamples;
        private final int timeoutSamples;
        private static final int NO_TIMEOUT = -1;
        private volatile boolean paused;
        private volatile boolean reset;
        RecognitionListener listener;

        public RecognizerThread(RecognitionListener listener, int timeout) {
            this.paused = false;
            this.reset = false;
            this.listener = listener;
            if (timeout != -1) {
                this.timeoutSamples = timeout * MySpeechService.this.sampleRate / 1000;
            } else {
                this.timeoutSamples = -1;
            }

            this.remainingSamples = this.timeoutSamples;
        }

        public RecognizerThread(RecognitionListener listener) {
            this(listener, -1);
        }

        public void setPause(boolean paused) {
            this.paused = paused;
        }

        public void reset() {
            this.reset = true;
        }

        public void run() {
            MySpeechService.this.recorder.startRecording();
            if (MySpeechService.this.recorder.getRecordingState() == 1) {
                MySpeechService.this.recorder.stop();
                IOException ioe = new IOException("Failed to start recording. Microphone might be already in use.");
                MySpeechService.this.mainHandler.post(() -> {
                    this.listener.onError(ioe);
                });
            }

            short[] buffer = new short[MySpeechService.this.bufferSize];

            while (!interrupted() && (this.timeoutSamples == -1 || this.remainingSamples > 0)) {
                int nread = MySpeechService.this.recorder.read(buffer, 0, buffer.length);
                if (!this.paused) {
                    if (this.reset) {
                        MySpeechService.this.recognizer.reset();
                        this.reset = false;
                    }

                    if (nread < 0) {
                        throw new RuntimeException("error reading audio buffer");
                    }

                    String result;
                    if (MySpeechService.this.recognizer.acceptWaveForm(buffer, nread)) {
                        result = MySpeechService.this.recognizer.getResult();
                        MySpeechService.this.mainHandler.post(() -> {
                            this.listener.onResult(result);
                        });
                    } else {
                        result = MySpeechService.this.recognizer.getPartialResult();
                        MySpeechService.this.mainHandler.post(() -> {
                            this.listener.onPartialResult(result);
                        });
                    }

                    if (this.timeoutSamples != -1) {
                        this.remainingSamples -= nread;
                    }
                }
            }

            MySpeechService.this.recorder.stop();
            if (!this.paused) {
                if (this.timeoutSamples != -1 && this.remainingSamples <= 0) {
                    MySpeechService.this.mainHandler.post(() -> {
                        this.listener.onTimeout();
                    });
                } else {
                    String finalResult = MySpeechService.this.recognizer.getFinalResult();
                    MySpeechService.this.mainHandler.post(() -> {
                        this.listener.onFinalResult(finalResult);
                    });
                }
            }

        }
    }
}

File: app/src/main/java/com/elishaazaria/sayboard/ime/MySpeechService.kt
/*
 * org.vosk.SpeechService, extended to support other recognizers.
 */
package com.elishaazaria.sayboard.ime

import android.Manifest
import android.media.AudioRecord
import android.media.MediaRecorder
import android.os.Handler
import android.os.Looper
import androidx.annotation.RequiresPermission
import com.elishaazaria.sayboard.ime.recognizers.Recognizer
import org.vosk.android.RecognitionListener
import java.io.IOException

class MySpeechService @RequiresPermission(Manifest.permission.RECORD_AUDIO) constructor(
    private val recognizer: Recognizer, sampleRate: Float
) {
    private val sampleRate: Int
    private val bufferSize: Int
    private val recorder: AudioRecord
    private var recognizerThread: RecognizerThread? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    init {
        this.sampleRate = sampleRate.toInt()
        bufferSize = Math.round(this.sampleRate.toFloat() * BUFFER_SIZE_SECONDS)
        recorder = AudioRecord(
            MediaRecorder.AudioSource.VOICE_RECOGNITION,
            this.sampleRate,
            16,
            2,
            bufferSize * 2
        )
        if (recorder.state == 0) {
            recorder.release()
            throw IOException("Failed to initialize recorder. Microphone might be already in use.")
        }
    }

    fun startListening(listener: RecognitionListener): Boolean {
        return if (null != recognizerThread) {
            false
        } else {
            recognizerThread =
                RecognizerThread(listener)
            recognizerThread!!.start()
            true
        }
    }

    fun startListening(listener: RecognitionListener, timeout: Int): Boolean {
        return if (null != recognizerThread) {
            false
        } else {
            recognizerThread =
                RecognizerThread(listener, timeout)
            recognizerThread!!.start()
            true
        }
    }

    private fun stopRecognizerThread(): Boolean {
        return if (null == recognizerThread) {
            false
        } else {
            try {
                recognizerThread!!.interrupt()
                recognizerThread!!.join()
            } catch (var2: InterruptedException) {
                Thread.currentThread().interrupt()
            }
            recognizerThread = null
            true
        }
    }

    fun stop(): Boolean {
        return stopRecognizerThread()
    }

    fun cancel(): Boolean {
        if (recognizerThread != null) {
            recognizerThread!!.setPause(true)
        }
        return stopRecognizerThread()
    }

    fun shutdown() {
        recorder.release()
    }

    fun setPause(paused: Boolean) {
        if (recognizerThread != null) {
            recognizerThread!!.setPause(paused)
        }
    }

    fun reset() {
        if (recognizerThread != null) {
            recognizerThread!!.reset()
        }
    }

    private inner class RecognizerThread @JvmOverloads constructor(
        var listener: RecognitionListener,
        timeout: Int = -1
    ) : Thread() {
        private var remainingSamples: Int
        private val timeoutSamples: Int

        @Volatile
        private var paused = false

        @Volatile
        private var reset = false

        init {
            if (timeout != -1) {
                timeoutSamples = timeout * sampleRate / 1000
            } else {
                timeoutSamples = -1
            }
            remainingSamples = timeoutSamples
        }

        fun setPause(paused: Boolean) {
            this.paused = paused
        }

        fun reset() {
            reset = true
        }

        override fun run() {
            recorder.startRecording()
            if (recorder.recordingState == 1) {
                recorder.stop()
                val ioe =
                    IOException("Failed to start recording. Microphone might be already in use.")
                mainHandler.post { listener.onError(ioe) }
            }
            val buffer = ShortArray(bufferSize)
            while (!interrupted() && (timeoutSamples == -1 || remainingSamples > 0)) {
                val nread = recorder.read(buffer, 0, buffer.size)
                if (!paused) {
                    if (reset) {
                        recognizer.reset()
                        reset = false
                    }
                    if (nread < 0) {
                        throw RuntimeException("error reading audio buffer")
                    }
                    var result: String?
                    if (recognizer.acceptWaveForm(buffer, nread)) {
                        result = recognizer.getResult()
                        mainHandler.post { listener.onResult(result) }
                    } else {
                        result = recognizer.getPartialResult()
                        mainHandler.post { listener.onPartialResult(result) }
                    }
                    if (timeoutSamples != -1) {
                        remainingSamples -= nread
                    }
                }
            }
            recorder.stop()
            if (!paused) {
                if (timeoutSamples != -1 && remainingSamples <= 0) {
                    mainHandler.post { listener.onTimeout() }
                } else {
                    val finalResult = recognizer.getFinalResult()
                    mainHandler.post { listener.onFinalResult(finalResult) }
                }
            }
        }
    }

    companion object {
        private const val NO_TIMEOUT = -1
        private const val BUFFER_SIZE_SECONDS = 0.2f
    }
}